### PR TITLE
Cleanup handling of overridden functions / constructors

### LINKED ISSUE #9: Cleanup method / factory overriding.

### PR DESCRIPTION
**Pull request by joshualitt**
* Adds support for optional parameters, and uses them to support overridden functions / constructors.
* Changes all uses of `extends` to `implements`. This is to avoid having to generate empty factory constructors.

Fixes #9

### CODE REVIEW

**Comment by joshualitt** [comment]
@kevmoo, we can start the review now, but this can't land until https://dart-review.googlesource.com/c/sdk/+/285701 lands.

**Comment by kevmoo** [comment]
@joshualitt – run the formatter quick!

File: tool/bindings_generator/translator.dart [hunk 1]
```diff
@@ -80,15 +80,145 @@ class _Library {
...
+      type = 'JSAny';
+    }
+    if (idlType.nullable.toDart) {
+      isNullable = true;
+    }
+  }
+}
+
+class _Parameter {
+  Set<String> _names;
```
Comment (kevmoo) [inline tool/bindings_generator/translator.dart:122]: could be final?

**Comment by joshualitt** [comment]
Lol, well, this passes but the behavior will be strange without the specialization CL, i.e. we'll pass `null` many places instead of nothing.

**Review by kevmoo — APPROVED** [review]
Did you consider updating the test file to use of these updated APIs?

**Comment by joshualitt** [comment]
> Did you consider updating the test file to use of these updated APIs?

I think we have to walk a fine line with testing in this package. We definitely need tests of the generator, I created #30 to track that effort. But I don't think we want to test the semantics of JS interop here, we should do that in the SDK. Nor do we want to test browser conformance.

That said, I'm not against more smoke tests. Let's think this through in #30.

**Pull request merged**